### PR TITLE
chore(facade): improve IoLoopV2 readability and simplify recv flow

### PR DESCRIPTION
**Pull request by romange**
## Summary

The main functional change - I pulled socket reading from `DoReadOnRecv` for the "regular", 
epoll-notification flow outside from the callback. Socket pending read state is separated from io_buf_ state by introducing an explicit `pending_input_` boolean.

Everything else is non-functional readability improvements.

- Introduce `HasInFlightCommands()` and `HasCommandToExecute()` predicates for the parsed command queue, replacing raw pointer comparisons
- Move socket draining from `DoReadOnRecv` into `IoLoopV2` directly via `pending_input_` flag, removing duplicated `TryRecv` logic
- Simplify `ReplyBatch` loop by advancing `parsed_head_` directly instead of aliasing through a reference variable with `exchange()`

## Changes

- **`src/facade/dragonfly_connection.h`** — Add `HasInFlightCommands()`, `HasCommandToExecute()` inline helpers; add `pending_input_` bit field
- **`src/facade/dragonfly_connection.cc`** — Refactor `IoLoopV2` recv loop to drain socket inline; simplify `ReplyBatch`; replace raw pointer comparisons with named predicates; simplify `DoReadOnRecv` to just set `pending_input_`; move `migration_allowed_to_register_` init into `IoLoopV2`
- **`tests/dragonfly/pymemcached_test.py`** — Add `test_get_many` for multi-key get coverage

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/facade/dragonfly_connection.h [hunk 1]
```diff
@@ -437,10 +437,23 @@ class Connection : public util::Connection {
...
   // Total bytes used by commands in parsed command queue
   size_t parsed_cmd_q_bytes_ = 0;
+
+  // Returns true if there are dispatched commands that haven't been replied yet.
+  bool HasInFlightCommands() const {
+    return parsed_head_ != parsed_to_execute_;
+  }
+
+  // Returns true if the head command is ready to execute (nothing in-flight ahead of it).
+  bool HasCommandToExecute() const {
```
Comment (glevkovich) [inline src/facade/dragonfly_connection.h]: I think that a more accurate name would be `IsHeadReadyToExecute()`. This is what is being checked here. It doesn't check if there is at least one command that can be executed, not necessarily the head:  it checks if the head command is the one currently waiting to be executed, ensuring no previously dispatched commands are blocking it.

File: src/facade/dragonfly_connection.h [hunk 2]
```diff
@@ -437,10 +437,23 @@ class Connection : public util::Connection {
...
   ParsedCommand* parsed_to_execute_ = nullptr;
+
   // Total number of commands in parsed command queue
   size_t parsed_cmd_q_len_ = 0;
+
   // Total bytes used by commands in parsed command queue
   size_t parsed_cmd_q_bytes_ = 0;
+
+  // Returns true if there are dispatched commands that haven't been replied yet.
+  bool HasInFlightCommands() const {
```
Comment (glevkovich) [inline src/facade/dragonfly_connection.h]: nit: suggest to rename HasInFlightCommands() to `HasDispatchedCommands().`
The term "in-flight" introduces new terminology and implies a specific execution state we can't strictly guarantee (they could be running or already completed). HasDispatchedCommands() is more accurate and grounded in the code's existing logic: we know the commands have been dispatched, even if we don't know their current processing status (running or completed and pending reply).

File: src/facade/dragonfly_connection.cc [hunk 3]
```diff
@@ -2712,33 +2682,46 @@ variant<error_code, Connection::ParserStatus> Connection::IoLoopV2() {
...
+            // TryRecv is non-blocking: it returns EAGAIN/EWOULDBLOCK when nothing is ready.
+            if (ec == errc::resource_unavailable_try_again || ec == errc::operation_would_block)
+              pending_input_ = false;
+            else
+              io_ec_ = ec;
+          }
+          break;
+        }
+        io_buf_.CommitWrite(*res);
+        buf = io_buf_.AppendBuffer();
```
Comment (dranikpg) [inline src/facade/dragonfly_connection.cc]: Please move such portions out of the functions, otherwise it becomes too large to be easily readable

**Comment by dranikpg** [comment]
oh, I merged my ioloop pr and now it has a conflict

File: src/facade/dragonfly_connection.cc [hunk 4]
```diff
@@ -2712,33 +2682,46 @@ variant<error_code, Connection::ParserStatus> Connection::IoLoopV2() {
...
+    }
+
+    if (pending_input_) {
+      // Drain available socket data into io_buf_.
+      io::MutableBytes buf = io_buf_.AppendBuffer();
+      // A recv call can return fewer bytes than requested even if the
+      // socket buffer actually contains enough data to satisfy the full request.
+      while (!buf.empty()) {
+        io::Result<size_t> res = socket_->TryRecv(buf);
+        if (!res || *res == 0) {
```
Comment (glevkovich) [inline src/facade/dragonfly_connection.cc]: I think you can simplify with a single layer of if/else (and 2 break):

```
io::Result<size_t> res = socket_->TryRecv(buf);
        
        if (!res) { //  / error
          auto ec = res.error();
          if (ec == errc::resource_unavailable_try_again || ec == errc::operation_would_block) {
            pending_input_ = false;
          } else {
            io_ec_ = ec;
          }
          break;
        } 
        else if (*res == 0) { // clean EOF
          io_ec_ = make_error_code(errc::connection_aborted);
          pending_input_ = false;
          break; 
        }
```

File: tests/dragonfly/pymemcached_test.py [hunk 5]
```diff
@@ -88,6 +88,12 @@ async def test_noreply_pipeline(self, df_server: DflyInstance, memcached_client:
 
         assert False, "Pipelining not detected"
 
+    def test_get_many(self, memcached_client: MCClient):
```
Comment (glevkovich) [inline tests/dragonfly/pymemcached_test.py:91]: AI generated:
The test_get_many method lacks coverage for edge cases such as empty keys list, non-existent keys, or mixed existing/non-existing keys. Add test cases to verify behavior with empty input and keys that don't exist in the cache.

and in more details:
test_get_many only tests the "happy path" (fetching 100 keys that all exist). In batch operations like Memcached's get (which supports multiple keys), it is standard practice to verify how the server handles missing keys. Memcached protocol specifically dictates that missing keys are simply omitted from the response, while existing ones are returned. Testing a mix of existing and non-existing keys ensures the parser and reply builder don't lose alignment or crash.

Comment (romange) [reply tests/dragonfly/pymemcached_test.py:91]: the goal is not to pursue correctness of `get` - it was to get more coverage for my changes as i had a bug and this test helped to debug it so I decided to keep it. The goal of this test is to increase coverage of facade.

File: src/facade/dragonfly_connection.h (hunk 1 above)
Comment (romange) [reply src/facade/dragonfly_connection.h]: I think I will push back on that as "head ready" is an implementation detail. What I want to know at high level is whether we have commands ready to dispatch. I will align it though with our jargon here and below

File: src/facade/dragonfly_connection.h (hunk 2 above)
Comment (romange) [reply src/facade/dragonfly_connection.h]: ok

File: src/facade/dragonfly_connection.h (hunk 1 above)
Comment (romange) [reply src/facade/dragonfly_connection.h]: Atually you are correct. We could execute other commands, so, if we have  `head-> ... -> to_execute-> tail` 
why do we call `ExecuteBatch` only if `head == to_execute` ? @dranikpg  what's the reason for that?

File: src/facade/dragonfly_connection.cc (hunk 3 above)
Comment (romange) [reply src/facade/dragonfly_connection.cc]: done

File: src/facade/dragonfly_connection.h (hunk 1 above)
Comment (dranikpg) [reply src/facade/dragonfly_connection.h]: It seems there is no good answer for it, it can be called eariler if there is something to execute

Comment (romange) [reply src/facade/dragonfly_connection.h]: @glevkovich  please add a TODO near the calls to this function to explore why we do not execute commands that are added to the queue while HasDispatchedCommands is true. Seems like an arbitrary decision.
